packages/components/src/ConfirmationModal/ConfirmationModal.tsx

```diff
@@ -0,0 +1,228 @@
+  forwardRef,
+  useEffect,
+  useImperativeHandle,
+  useReducer,
+} from "react";
+import { Text } from "../Text";
+import { Modal } from "../Modal";
+import { Content } from "../Content";
+
+interface ConfirmationModalState {
+  readonly title?: string;
+  readonly message?: string;
+  readonly open: boolean;
+  readonly confirmLabel: string;
+  readonly cancelLabel: string;
+  onConfirm?(): void;
+  onCancel?(): void;
+}
+
+interface ConfirmOrCancelAction {
+  type: "confirm" | "cancel";
+}
+
+interface DisplayAction {
+  type: "display";
+  title?: string;
+  message: string;
+  confirmLabel: string;
+  cancelLabel?: string;
+  onConfirm(): void;
+  onCancel?(): void;
+}
+
+interface ResetAction {
+  type: "reset";
+  state: ConfirmationModalState;
+}
+
+function confirmationModalReducer(
+  state: ConfirmationModalState,
+  action: ConfirmOrCancelAction | ResetAction | DisplayAction,
+) {
+  switch (action.type) {
+    case "display":
+      return {
+        ...state,
+        title: action.title,
+        message: action.message,
+        open: true,
+        confirmLabel: action.confirmLabel,
+        cancelLabel: action.cancelLabel || "Cancel",
+        onConfirm: action.onConfirm,
+        onCancel: action.onCancel,
+      };
+
+    case "confirm":
+      state.onConfirm && state.onConfirm();
+      return {
+        ...state,
+        open: false,
+      };
+
+    case "cancel":
+      state.onCancel && state.onCancel();
+      return {
+        ...state,
+        open: false,
+      };
+
+    case "reset":
+      return { ...state, ...action.state };
+
+    default:
+      throw new Error();
+  }
+}
+
+export interface ConfirmationModalRef {
+  show(props: Omit<DisplayAction, "type">): void;
+}
+
+interface BaseConfirmationModalProps {
+  /**
+   * Title for the modal.
+   */
+  readonly title?: string;
+
+  /**
+   * Text for the body of the modal.
+   */
+  readonly message?: string;
+
+  /**
+   * Controls if the modal is open or not.
+   */
+  readonly open?: boolean;
+
+  /**
+   * Label for the confirm button.
+   */
+  readonly confirmLabel?: string;
+
+  /**
+   * Label for the cancel button.
+   */
+  readonly cancelLabel?: string;
+
+  /**
+   * Callback for when the confirm button is pressed.
+   */
+  onConfirm?(): void;
+
+  /**
+   * Callback for when the cancel button is pressed.
+   */
+  onCancel?(): void;
+
+  /**
+   * Callback for whenever a user's action should close the modal.
+   */
+  onRequestClose?(): void;
+}
+
+interface SimpleConfirmationModalProps extends BaseConfirmationModalProps {
+  readonly message: string;
+  readonly open: boolean;
+  readonly confirmLabel: string;
+}
+
+interface ComplexConfirmationModalProps extends BaseConfirmationModalProps {
+  readonly ref: Ref<ConfirmationModalRef>;
+  readonly open?: undefined;
+}
+
+type ConfirmationModalProps =
+  | SimpleConfirmationModalProps
+  | ComplexConfirmationModalProps;
+
+export const ConfirmationModal = forwardRef(ConfirmationModalInternal);
```

Comment:
for OCD reasons, can we move this to the bottom of the file?

Reply:
Talked through an even better way!